File: examples/word-add-revision.ts
/**
 * 例子（Epic-19）：构造一份 docx，原段 "Hello old world"，把 "old " 标记为删除、
 * "kind " 标记为插入，演示 Word 修订追踪。
 *
 * 跑法：
 *   bun run examples/word-add-revision.ts <output.docx>
 */

import {
  Paragraph,
  Run,
  Text,
  WordprocessingDocument,
  createDeletedRun,
  createInsertedRun,
} from "../src/word/index.js";

async function main(): Promise<void> {
  const [outputPath] = process.argv.slice(2);
  if (outputPath === undefined) {
    process.stderr.write("usage: word-add-revision <output.docx>\n");
    process.exit(2);
  }

  const doc = WordprocessingDocument.create();
  const body = doc.mainDocumentPart?.document.firstChild()!;
  const append = (e: unknown): void =>
    (body as { appendChild: (e: unknown) => void }).appendChild(e);

  const p = new Paragraph();

  // 前置纯文本
  const lead = new Run();
  const leadT = new Text();
  leadT.text = "Hello ";
  lead.appendChild(leadT);
  p.appendChild(lead);

  // 删除 "old " + 插入 "kind "（Alice 改的）
  const author = "Alice";
  const date = new Date().toISOString();
  p.appendChild(createDeletedRun({ author, id: doc.nextRevisionId(), date, text: "old " }));
  p.appendChild(createInsertedRun({ author, id: doc.nextRevisionId(), date, text: "kind " }));

  // 尾部纯文本
  const tail = new Run();
  const tailT = new Text();
  tailT.text = "world";
  tail.appendChild(tailT);
  p.appendChild(tail);

  append(p);

  await doc.saveAsAsync(outputPath);
  process.stdout.write(`Wrote ${outputPath} (1 del + 1 ins by Alice)\n`);
}

main().catch((err) => {
  process.stderr.write(`failed: ${(err as Error).stack ?? err}\n`);
  process.exit(1);
});

File: tools/cross-sdk-verify/replicas/WordAddRevision.cs
// Replica of examples/word-add-revision.ts
// Creates a docx with revision tracking:
//   "Hello " + deleted "old " + inserted "kind " + "world"
// Author: Alice

using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;

namespace CrossSdkVerify.Replicas;

public static class WordAddRevision
{
    public static void Run(string outputPath)
    {
        using var doc = WordprocessingDocument.Create(outputPath,
            DocumentFormat.OpenXml.WordprocessingDocumentType.Document);

        var mainPart = doc.AddMainDocumentPart();
        mainPart.Document = new Document();
        var body = new Body();
        mainPart.Document.Append(body);

        // Fixed date for reproducibility
        var date = new DateTime(2024, 1, 1, 0, 0, 0, DateTimeKind.Utc);
        const string author = "Alice";

        var p = new Paragraph();

        // "Hello "
        p.Append(new Run(new Text("Hello ") { Space = SpaceProcessingModeValues.Preserve }));

        // Deleted "old "
        var del = new DeletedRun(
            new RunProperties(),
            new DeletedText("old ") { Space = SpaceProcessingModeValues.Preserve })
        {
            Id = "1",
            Author = author,
            Date = date
        };
        p.Append(del);

        // Inserted "kind "
        var ins = new InsertedRun(
            new Run(new Text("kind ") { Space = SpaceProcessingModeValues.Preserve }))
        {
            Id = "2",
            Author = author,
            Date = date
        };
        p.Append(ins);

        // "world"
        p.Append(new Run(new Text("world")));

        body.Append(p);
        body.Append(new SectionProperties());
        mainPart.Document.Save();
    }
}
